checker/src/org/checkerframework/checker/index/searchindex/SearchIndexAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,117 @@
+
+/**
+ * The Search Index Checker is used to help type the results of calls to the JDK's binary search
+ * methods. It is part of the Index Checker.
+ */
+public class SearchIndexAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    public final AnnotationMirror UNKNOWN, BOTTOM;
+
+    public SearchIndexAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        UNKNOWN = AnnotationUtils.fromClass(elements, SearchIndexUnknown.class);
+        BOTTOM = AnnotationUtils.fromClass(elements, SearchIndexBottom.class);
+        this.postInit();
+    }
+
+    /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant prop + folding).
+     */
+    ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
+        return getTypeFactoryOfSubchecker(ValueChecker.class);
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return new LinkedHashSet<>(
+                Arrays.asList(
+                        SearchIndex.class,
+                        SearchIndexBottom.class,
+                        SearchIndexUnknown.class,
+                        NegativeIndexFor.class));
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(
+            MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+        return new SearchIndexQualifierHierarchy(factory);
+    }
+
+    private final class SearchIndexQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public SearchIndexQualifierHierarchy(
+                MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexUnknown.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexBottom.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexUnknown.class)) {
+                return false;
+            }
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexBottom.class)) {
+                return false;
+            }
+            // All annotations have arguments, because each is either NegativeIndexFor or
+            // SearchIndex. All NegativeIndexFor annotations are subtypes of SearchIndex
+            // annotations whose arguments contain their arguments.
+            if (AnnotationUtils.areSameByClass(subAnno, NegativeIndexFor.class)) {
+                List<String> superArrays =
+                        IndexUtil.getValueOfAnnotationWithStringArgument(superAnno);
+                List<String> subArrays = IndexUtil.getValueOfAnnotationWithStringArgument(subAnno);
+                return superArrays != null
```

Comment:
If superArrays or subArrays is null, is that an indication of a bug?  If so, it's better to check it and throw an exception than mask it by returning a value.

You could set superArrays and subArrays outside the if statements so the code is shorter and only has to check for nullness once.